File: 303. Range Sum Query - Immutable/range-sum-query-immutable.java
class NumArray {
    static class SegmentTree {
        int leftmost, rightmost;
        int sum;
        SegmentTree lChild, rChild;
        public SegmentTree(int leftmost, int rightmost, int[] a) {
            this.leftmost = leftmost;
            this.rightmost = rightmost;
            if (leftmost == rightmost) {
                sum = a[leftmost];
            } else {
                int mid = (leftmost + rightmost) / 2;
                lChild = new SegmentTree(leftmost, mid, a);
                rChild = new SegmentTree(mid + 1, rightmost, a);
                recalc();
            }
        }
        public void recalc(){
            if (leftmost == rightmost) return;
            sum = lChild.sum + rChild.sum;
        }
        public void pointUpdate(int index, int newVal) {
            if (leftmost == rightmost) {
                sum = newVal;
                return;
            }
            if (index <= lChild.rightmost) lChild.pointUpdate(index, newVal);
            else rChild.pointUpdate(index, newVal);
            recalc();
        }
        public int rangeSum(int l, int r) {
            if (l > rightmost || r < leftmost) {
                return 0;
            } else if (l <= leftmost && r >= rightmost) {
                return sum;
            }
            return lChild.rangeSum(l, r) + rChild.rangeSum(l, r);
        }
    }

    private SegmentTree st;
    public NumArray(int[] nums) {
        st = new SegmentTree(0, nums.length - 1, nums);
    }
    
    public int sumRange(int left, int right) {
        return st.rangeSum(left, right);
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(left,right);
 */

File: 303. Range Sum Query - Immutable/range-sum-query-immutable.py
# Since the array is immutable
# A segment tree isn't optimal
# A simple prefix sum is enough 
class NumArray:
    def __init__(self, nums: List[int]):
        # O(n)
        self.pref = [0]
        for num in nums:
            self.pref.append(self.pref[-1] + num)

    def sumRange(self, left: int, right: int) -> int:
        # O(1)
        return self.pref[right + 1] - self.pref[left]


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)
